In [1]:
from pulp import LpVariable, LpProblem, LpMaximize, LpStatus, value, LpMinimize, GLPK

# Question 1: 
__1A.  Write the general dual problem associated with the given LP__    
(Do not transform or rewrite the primal problem before writing the general dual) 
 
__Maximize__ $$–4x_1 + 2x_2$$
__Subject To__
$$4x_1 + x_2 + x_3 = 20$$
$$2x_1 – x_2 ≥ 6$$
$$x_1 – x_2 + 5x_3 ≥ –5$$
$$–3x_1 + 2x_2 + x_3 ≤ 4$$ 

$$x_1 ≤ 0, x_2 ≥ 0, x_3\text{ unrestricted}$$


## Answer:

__Minimize__ $$20y_1 + 6y_2 - 5y_3 + 4y_4$$
__Subject To__
$$4y_1 + 2y_2 + y_3 - 3y_4 ≤ -4$$
$$y_1 - y_2 - y_3 + 2y_4 ≥ 3$$
$$y_1 + 5y_3 + y_4 = 0$$

$$y_1\text{ unrestricted}, y_2 ≤ 0, y_3 ≤ 0, y_4 ≥ 0$$


1B.  Given the following information for a product-mix problem with three products and three resources. 
Primal Decision Variables:  
* x1 = number of unit 1 produced
* x2 = # of unit 2 produced
* x3 = # of unit 3 produced 

__Primal Formulation:__       
$$\text{Max Z (Rev.) } =  25x1  + 30x2   + 20x3$$     
Subject To 
$$8x1 + 6x2 + x3 ≤ 50 \text{ (Res. 1 constraint)}$$ 
$$4x1 + 2x2 + 3x3 ≤ 20 \text{ (Res. 2 constraint)}$$   
$$2x1 + x2 + 2x3 ≤ 25 \text{ (Res. 3 constraint)}$$   
$$x1, x2, x3 ≥ 0  \text{( Nonnegativity)}$$    

__Dual Formulation:__
$$\text{Min W} =  50π1 + 20π2 +25π3$$ 
Subject To    
$$8π1 + 4π2 +2π3≥ 25$$
$$6π1 + 2π2 +π3  ≥ 30$$ 
$$π1 + 3π2 +2π3≥ 20$$ 
$$π1, π2, π3          ≥ 0$$ 

__Optimal Solution:__      
Optimal Z = Revenue = \\$268.75   
$$x1 = 0 \text{ (Number of unit 1)}$$  
$$x2 = 8.125 \text{ (Number of unit 2)}$$   
$$x3 = 1.25 \text{ (Number of unit 3)}$$   

Dual Var. Optimal Value = 22.5 (Surplus variable in 1st dual constraint)   
Dual Var. Optimal Value = 0 (Surplus variable in 2nd dual constraint)   
Dual Var. Optimal Value = 0 (Surplus variable in 3rd dual constraint)  

__Resource Constraints:__    
Resource 1 = 0 leftover units    
Resource 2 = 0 leftover units   
Resource 3 = 14.375 leftover units   

Dual Var. Optimal Value = 3.125 = π1   
Dual Var. Optimal Value = 5.625 = π2   
Dual Var. Optimal Value = 0 = π3    
 
## __1Bi. What is the fair-market price for one unit of Resource 3?__   
\\$0.  This is because the optimal solution of the dual is the vector of shadow prices (https://web.stanford.edu/~ashishg/msande111/notes/chapter4.pdf).  In this case, the shadow price is 0, so the company would not be willing to pay anything for one unit of resource 3 since it will not impact the outcome - the optimized revenue of \\268.75 would not change. 
 
## __1Bii. What is the meaning of the surplus variable value of 22.5 in the 1st dual constraint with respect to the primal problem?__

This is the equivalent of the shadow price of unit 1 from the primal - the optimal solution to the dual provides shadow prices.

# Question 2: 

Carco manufactures cars and trucks.  Each car contributes \\$300 to profit and each truck, \\$400; these profits do not consider machine rental.  The resources required to manufacture a car and a truck are shown below.  Each day 
Carco can rent up to 98 Type 1 machines at a cost of \\$50 per machine.  The company now has 73 Type 2 machines 
and 260 tons of steel available.  Marketing considerations dictate that at least 88 cars and at least 26 trucks be 
produced. 

|Vehicle Type  |Days On Machine 1|Days on Machine 2|Tons of Steel  |
|-----|---|---|---|
|Car  |0.8|0.6|2  |
|Truck|1  |0.7|3  |
 
__Part A:  Formulate the problem as a Linear Program.__  

__Part B:  Solve the LP (provide exact values for all variables and the optimal objective function).__    
Hint:  The optimal objective function value is $32540 

In [2]:
# Problem 2
# define variables
cars = LpVariable("cars", 0, None)
trucks = LpVariable("trucks", 0, None)
type_1_machines = LpVariable("type_1_machines", 0, None)

# defines the problem
prob2 = LpProblem("CarCo Problem", LpMaximize)

# define objective function - Maximize profit
prob2 += 300*cars + 400*trucks - (type_1_machines*50)

# define constraints
prob2 += (.8*cars + trucks - type_1_machines <= 0, 'Cars and trucks take up type 1 machine time')
prob2 += (type_1_machines <= 98, 'Carco can rent up to 98 Type 1 machines')
prob2 += (.6*cars + .7*trucks<= 73, 'Carco now has 73 Type 2 machines')
prob2 += (2*cars + 3*trucks <= 260, 'The company has 260 tons of steel available')
prob2 += (cars >= 88, 'Marketing considerations dictate that at least 88 cars be produced')
prob2 += (trucks >= 26, 'Marketing considerations dictate that at least 26 trucks be produced')

# solve the problem
prob2.writeLP("prob2.lp")
prob2.solve(GLPK(msg=0, options=['--ranges', 'prob2.sen']))
print ("Status:", LpStatus[prob2.status])

# Note, we are only able to get sensitivity information because we are solving
# as a linear program.  If we solved as an Integer Program, then no 
# sensitivity information would be available.
print ("Objective", value(prob2.objective))
for v in prob2.variables():
    print(v.name, "=", v.varValue)
print ("")

f = open("prob2.sen", "r")
print(f.read())

Status: Optimal
Objective 32540.0
cars = 88.0
trucks = 27.6
type_1_machines = 98.0

GLPK 5.0  - SENSITIVITY ANALYSIS REPORT                                                                         Page   1

Problem:    
Objective:  OBJ = 32540 (MAXimum)

   No. Row name     St      Activity         Slack   Lower bound       Activity      Obj coef  Obj value at Limiting
                                          Marginal   Upper bound          range         range   break point variable
------ ------------ -- ------------- ------------- -------------  ------------- ------------- ------------- ------------
     1 Carco_can_rent_up_to_98_Type_1_machines
                    NU      98.00000        .               -Inf       96.40000    -350.00000   31980.00000 Marketing_considerations_dictate_that_at_least_26_trucks_be_produced
                                         350.00000      98.00000       98.40000          +Inf   32680.00000 The_company_has_260_tons_of_steel_available

     2 Carco_n

/Users/robords/.venv/basic_linear_programming/lib/python3.9/site-packages/pulp/pulp.py:1352: UserWarning: Spaces are not permitted in the name. Converted to '_'
  warnings.warn("Spaces are not permitted in the name. Converted to '_'")


__Part C:  Answer the following questions from your output.  (Note, do not simply rerun the model – use the Linear 
Programming output and Sensitivity Analysis to explain your answers.)__     
i)  If cars contributed \\$310 to profit, what would be the new optimal solution to the problem?     
* Since the objective range coefficient is between 0 and 320, in this case, increasing it to 310 will not change the number of cars or trucks produced, it will simply increase the profit.  So, since we are producing 88 cars in this model and we increased car profit by $10, 88 * 10 = 880, so the new profit/optimal solution would be \\$32540 + \\$880 = __\\$33420__

ii) What is the most that Carco should be willing to pay to rent an additional Type 1 machine for 1 day?    
\\$350, since the shadow price in the sensitivity report provides this.  This means that any additional unit for one day of Type 1 machines will result in an increase optimal value of \\$350.  This is only valid for one additional unit.

iii) What is the most that Carco should be willing to pay for an extra ton of steel?   
Nothing, since it will not impact the optimum value - the shadow price for steel is null.

iv) If Carco were required to produce at least 86 cars, what would Carco’s profit become?    
It would increase by \\$40 to \\$32580, since we would produce more trucks.  Trucks are more profitable.  This is shown in the `Marketing_considerations_dictate_that_at_least_88_cars_be_produced` constraint, where each unit decrease for a car in this constraint, increases the objective by \\$20.

v) Carco is considering producing jeeps.  A jeep contributes \\$600 to profit and requires 1.2 days on machine 1, 2 
days on machine 2, and 4 tons of steel.  Should Carco produce any jeeps?  

No - the machine hours are too "expensive" in this case, as can be seen in the dual

## Question 3: 

A catering company must have the following number of clean napkins available at the beginning of each of the next 
four days:  day 1: 15, day 2: 12, day 3: 18, and day 4: 6.  After being used, a napkin can be cleaned by one of two 
methods: fast service or slow service. Fast service costs \\$0.10 per napkin, and a napkin cleaned via fast service is 
available for use the day after it is last used.  Slow service costs \\$0.06 per napkin, and a napkin cleaned via slow 
service is available two days after they were last used.  New napkins can be purchased for a cost of \\$0.20 per 
napkin. 
 
__Part A:  Formulate the problem as a minimum cost transportation problem.__     
__Part B:  Solve the problem (provide exact values for all variables and the optimal objective function).__

In [10]:
# Problem 3
# define variables
day1_new_nap = LpVariable("new_napkins_purchased_to_day_1", 0, None, cat='Integer')
day2_new_nap = LpVariable("new_napkins_purchased_to_day_2", 0, None, cat='Integer')
day3_new_nap = LpVariable("new_napkins_purchased_to_day_3", 0, None, cat='Integer')
day4_new_nap = LpVariable("new_napkins_purchased_to_day_4", 0, None, cat='Integer')

day1_fast_to_2 = LpVariable("day1_fast_to_day_2", 0, None, cat='Integer')
day1_fast_to_3 = LpVariable("day1_fast_to_day_3", 0, None, cat='Integer')
day1_fast_to_4 = LpVariable("day1_fast_to_day_4", 0, None, cat='Integer')

day2_fast_to_3 = LpVariable("day2_fast_to_day_3", 0, None, cat='Integer')
day2_fast_to_4 = LpVariable("day2_fast_to_day_4", 0, None, cat='Integer')

day3_fast_to_4 = LpVariable("day3_fast_to_day_4", 0, None, cat='Integer')

day1_slow_to_3 = LpVariable("day1_slow_to_day_3", 0, None, cat='Integer')
day1_slow_to_4 = LpVariable("day1_slow_to_day_4", 0, None, cat='Integer')

day2_slow_to_4 = LpVariable("day2_slow_to_day_4", 0, None, cat='Integer')



# defines the problem
napkin_prob = LpProblem("Napkin Problem", LpMinimize)

# define objective function - Maximize profit
napkin_prob += (.1*(day1_fast_to_2 + day1_fast_to_3 + day1_fast_to_4 + day2_fast_to_3 
                    + day2_fast_to_4 + day3_fast_to_4) + 
                .06*(day1_slow_to_3 + day1_slow_to_4 + day2_slow_to_4) + 
                .2*(day1_new_nap + day2_new_nap + day3_new_nap + day4_new_nap))

# define constraints
## We must have the following number of clean napkins available at the beginning of each of the next four days
napkin_prob += (day1_new_nap >= 15, 'day 1 15') ## assumes we have to buy all new napkins to start
napkin_prob += (day1_fast_to_2 + day2_new_nap >= 12, 'day 2 12')
napkin_prob += (day1_fast_to_3 + day2_fast_to_3 + day1_slow_to_3 +  day3_new_nap >= 18, 'day 3 18')
napkin_prob += (day1_fast_to_4 + day2_fast_to_4 + day3_fast_to_4 + day1_slow_to_4 + 
                day2_slow_to_4 >= 6, 'day 4 6')

napkin_prob += (day1_fast_to_2 + day1_fast_to_3 + day1_fast_to_4 + 
                day1_slow_to_3  + day1_slow_to_4 <= day1_new_nap, 'day 1 napkins must exist') 
napkin_prob += (day2_new_nap - day2_fast_to_3 - day2_fast_to_4 - day2_slow_to_4
                >= 0, 'day 2 napkins must exist') 
napkin_prob += (day3_new_nap - day3_fast_to_4 >= 0, 'day 3 napkins must exist') 

# solve the problem
napkin_prob.writeLP("napkin_prob.lp")
napkin_prob.solve(GLPK(msg=0))
print ("Status:", LpStatus[napkin_prob.status])

# Note, we are only able to get sensitivity information because we are solving
# as a linear program.  If we solved as an Integer Program, then no 
# sensitivity information would be available.
print ("Objective", value(napkin_prob.objective))
for v in napkin_prob.variables():
    print(v.name, "=", v.varValue)
print ("")

Status: Optimal
Objective 6.8999999999999995
day1_fast_to_day_2 = 1
day1_fast_to_day_3 = 0
day1_fast_to_day_4 = 0
day1_slow_to_day_3 = 14
day1_slow_to_day_4 = 0
day2_fast_to_day_3 = 4
day2_fast_to_day_4 = 0
day2_slow_to_day_4 = 6
day3_fast_to_day_4 = 0
new_napkins_purchased_to_day_1 = 15
new_napkins_purchased_to_day_2 = 11
new_napkins_purchased_to_day_3 = 0
new_napkins_purchased_to_day_4 = 0



/Users/robords/.venv/basic_linear_programming/lib/python3.9/site-packages/pulp/pulp.py:1352: UserWarning: Spaces are not permitted in the name. Converted to '_'
  warnings.warn("Spaces are not permitted in the name. Converted to '_'")


## Question 4: 

A university has three professors who each teach four courses per year.  Each year, four sections of marketing, 
finance, and production must be offered.  At least one section of each class must be offered during each semester 
(fall and spring).  Each professor’s time preferences and preference for teaching various courses are given below. 
 
The total satisfaction a professor earns teaching a class is the sum of the semester satisfaction and the course 
satisfaction.  Thus, professor 1 derives a satisfaction of 3 + 6 = 9 from teaching marketing during the fall semester. 
 
Part A:  Formulate the problem as a minimum cost network flow problem that can be used to assign professors to 
courses so as to maximize the total satisfaction of the three professors.  Draw the network and identify the nodes 
and arcs. 

|                   |Professor 1|Professor 2|Professor 3|
|-------------------|-----------|-----------|-----------|
|Fall Preference    |3          |5          |4          |
|Spring Preference  |4          |3          |4          |
|Marketing          |6          |4          |5          |
|Finance            |5          |6          |4          |
|Production         |4          |5          |6          |

![alt text](Homework2_Professor_Problem-Professor_Problem.drawio.png "BMC")
 
Part B:  Solve the problem (provide exact values for all variables and the optimal objective function). 


In [20]:
# Problem 4
# define variables - each arc is a variable

P1_Fall_Marketing = LpVariable("Professor 1 Fall Marketing", 0, None, cat='Integer')  
P1_Fall_Finance = LpVariable("Professor 1 Fall Finance", 0, None, cat='Integer')
P1_Fall_Production = LpVariable("Professor 1 Fall Production", 0, None, cat='Integer')
P1_Spring_Marketing = LpVariable("Professor 1 Spring Marketing", 0, None, cat='Integer')
P1_Spring_Finance = LpVariable("Professor 1 Spring Finance", 0, None, cat='Integer')
P1_Spring_Production = LpVariable("Professor 1 Spring Production", 0, None, cat='Integer')

P2_Fall_Marketing = LpVariable("Professor 2 Fall Marketing", 0, None, cat='Integer')  
P2_Fall_Finance = LpVariable("Professor 2 Fall Finance", 0, None, cat='Integer')
P2_Fall_Production = LpVariable("Professor 2 Fall Production", 0, None, cat='Integer')
P2_Spring_Marketing = LpVariable("Professor 2 Spring Marketing", 0, None, cat='Integer')
P2_Spring_Finance = LpVariable("Professor 2 Spring Finance", 0, None, cat='Integer')
P2_Spring_Production = LpVariable("Professor 2 Spring Production", 0, None, cat='Integer')

P3_Fall_Marketing = LpVariable("Professor 3 Fall Marketing", 0, None, cat='Integer')  
P3_Fall_Finance = LpVariable("Professor 3 Fall Finance", 0, None, cat='Integer')
P3_Fall_Production = LpVariable("Professor 3 Fall Production", 0, None, cat='Integer')
P3_Spring_Marketing = LpVariable("Professor 3 Spring Marketing", 0, None, cat='Integer')
P3_Spring_Finance = LpVariable("Professor 3 Spring Finance", 0, None, cat='Integer')
P3_Spring_Production = LpVariable("Professor 3 Spring Production", 0, None, cat='Integer')


# defines the problem
professor_problem = LpProblem("Professor Problem", LpMaximize)

# define objective function - Maximize Satisfaction

professor_problem += ((3+6)*P1_Fall_Marketing + (3+5)*P1_Fall_Finance + (3+4)*P1_Fall_Production +
                      (4+6)*P1_Spring_Marketing + (4+5)*P1_Spring_Finance + (4+4)*P1_Spring_Production +
                      (5+4)*P2_Fall_Marketing + (5+6)*P2_Fall_Finance + (5+5)*P2_Fall_Production +
                      (3+4)*P2_Spring_Marketing + (3+6)*P1_Spring_Finance + (3+5)*P2_Spring_Production +
                      (4+5)*P3_Fall_Marketing + (4+4)*P3_Fall_Finance + (4+6)*P3_Fall_Production +
                      (4+5)*P3_Spring_Marketing + (4+4)*P3_Spring_Finance + (4+6)*P3_Spring_Production)
                      
# define constraints - Each Node is a constraint

## A university has three professors who each teach four courses per year.
professor_problem += (-P1_Fall_Marketing - P1_Fall_Finance - P1_Fall_Production -
                      P1_Spring_Marketing - P1_Spring_Finance - P1_Spring_Production == -4, 
                      'Professor 1 teaches four courses a year') 
professor_problem += (-P2_Fall_Marketing - P2_Fall_Finance - P2_Fall_Production -
                      P2_Spring_Marketing - P2_Spring_Finance - P2_Spring_Production == -4, 
                      'Professor 2 teaches four courses a year') 
professor_problem += (-P3_Fall_Marketing - P3_Fall_Finance - P3_Fall_Production -
                      P3_Spring_Marketing - P3_Spring_Finance - P3_Spring_Production == -4, 
                      'Professor 3 teaches four courses a year')  

# At least one section of each class must be offered during each semester (fall and spring).
professor_problem += (P1_Fall_Marketing + P2_Fall_Marketing + P3_Fall_Marketing >= 1,
                      'One section of Marketing every fall')
professor_problem += (P1_Fall_Finance + P2_Fall_Finance + P3_Fall_Finance >= 1,
                      'One section of Finance per every fall')
professor_problem += (P1_Fall_Production + P2_Fall_Production + P3_Fall_Production >= 1,
                      'One section of Production per every fall')

professor_problem += (P1_Spring_Marketing + P2_Spring_Marketing + P3_Spring_Marketing >= 1,
                      'One section of Marketing every spring')
professor_problem += (P1_Spring_Finance + P2_Spring_Finance + P3_Spring_Finance >= 1,
                      'One section of Finance per every spring')
professor_problem += (P1_Spring_Production + P2_Spring_Production + P3_Spring_Production >= 1,
                      'One section of Production per every spring')

# Each year, four sections of marketing, finance, and production must be offered
professor_problem += (P1_Fall_Marketing + P1_Spring_Marketing + P2_Fall_Marketing + 
                      P2_Spring_Marketing + P3_Fall_Marketing + P3_Spring_Marketing >= 4,
                      'Four sections of marketing per year')
professor_problem += (P1_Fall_Finance + P1_Spring_Finance + P2_Fall_Finance + 
                      P1_Spring_Finance + P3_Fall_Finance + P3_Spring_Finance >= 4,
                      'Four sections of finance per year')
professor_problem += (P1_Fall_Production + P1_Spring_Production + P2_Fall_Production + 
                      P2_Spring_Production + P3_Fall_Production + P3_Spring_Production >= 4,
                      'Four sections of production per year')


# solve the problem
professor_problem.writeLP("professor_problem.lp")
professor_problem.solve(GLPK(msg=0))
print ("Status:", LpStatus[professor_problem.status])

# Note, we are only able to get sensitivity information because we are solving
# as a linear program.  If we solved as an Integer Program, then no 
# sensitivity information would be available.
print ("Objective", value(professor_problem.objective))
for v in professor_problem.variables():
    print(v.name, "=", v.varValue)
print ("")


Status: Optimal
Objective 142
Professor_1_Fall_Finance = 0
Professor_1_Fall_Marketing = 0
Professor_1_Fall_Production = 0
Professor_1_Spring_Finance = 3
Professor_1_Spring_Marketing = 1
Professor_1_Spring_Production = 0
Professor_2_Fall_Finance = 1
Professor_2_Fall_Marketing = 2
Professor_2_Fall_Production = 1
Professor_2_Spring_Finance = 0
Professor_2_Spring_Marketing = 0
Professor_2_Spring_Production = 0
Professor_3_Fall_Finance = 0
Professor_3_Fall_Marketing = 0
Professor_3_Fall_Production = 0
Professor_3_Spring_Finance = 0
Professor_3_Spring_Marketing = 1
Professor_3_Spring_Production = 3



/Users/robords/.venv/basic_linear_programming/lib/python3.9/site-packages/pulp/pulp.py:1352: UserWarning: Spaces are not permitted in the name. Converted to '_'
  warnings.warn("Spaces are not permitted in the name. Converted to '_'")
